tests/PerformanceTest.py

```diff
@@ -0,0 +1,509 @@
+
+                # Store rule-specific results for creating separate sheets in Excel
+                if rule_name not in rule_results:
+                    rule_results[rule_name] = []
+                rule_results[rule_name].append(
+                    {
+                        "Dataset": dataset_name,
+                        "Number of Calls": rule_executions,
+                        "Mean Time": sum(time_taken) / len(time_taken),
+                        "Median Time": median(time_taken),
+                        "Min Time": min(time_taken),
+                        "Max Time": max(time_taken),
+                        "Preprocessing Time": ", ".join(map(str, preprocessing_times)),
+                        "Operator Times": ", ".join(
+                            [
+                                f"{op}: {durations}"
+                                for op, durations in rule_operator_times.items()
+                            ]
+                        ),
+                        "Operation Times": ", ".join(map(str, all_operation_times)),
+                    }
+                )
+
+        # After all rules have been processed for a dataset, calculate the overall stats
+        if all_time_taken:
+            results.append(
+                {
+                    "function type": "TimeTestFunction",
+                    "rule name": "All Rules Combined",
+                    "dataset": dataset_name,
+                    "status": "Successful",
+                    "Number of Calls for each rule": total_calls,
+                    "Mean Time": sum(all_time_taken) / len(all_time_taken),
+                    "Median Time": median(all_time_taken),
+                    "Min Time": min(all_time_taken),
+                    "Max Time": max(all_time_taken),
+                    "Preprocessing Time": ", ".join(map(str, all_preprocessing_times)),
+                    "Operator Times": all_operator_times,
+                    "Operation Times": ", ".join(map(str, all_operation_times)),
+                    "Error": None,
+                }
+            )
+
+    return results, rule_results
+
+
+def all_datset_against_each_rule(dataset_dir, rule_dir, total_calls):
+    results = []  # To store the final report
+    dataset_results = {}  # To store dataset-specific results for Excel sheet creation
+
+    dataset_files = [
+        os.path.join(dataset_dir, file)
+        for file in os.listdir(dataset_dir)
+        if file.endswith((".json", ".xpt"))
+    ]
+    rules = [
+        file
+        for file in os.listdir(rule_dir)
+        if os.path.isfile(os.path.join(rule_dir, file))
+    ]
+
+    for rule in rules:
+        rule_name = os.path.basename(rule)
+
+        # Initialize variables to collect times for the dataset across all rules
+        all_time_taken = []
+        all_preprocessing_times = []
+        all_operator_times = {}
+        all_operation_times = []
+
+        rule_names = []
+        for dataset_path in dataset_files:

```

Comment:
It looks like this method is essentially running the same set of commands as the previous function, just in a different order. I think it is redundant.
I don't think it supports rules where the rule might require multiple datasets that are joined together since only a single dataset is being passed at a time. I think the timing feedback needs to given from within the engine. Maybe an additional logging mechanism can be used.